Updated system-nixpkgs-map.nix and a default for X11 package (fixes #984)

paumr opened the pull request:
This argument allows to pass a set of additional package-mappings as a
parameter rather than modifying them globally.

This seems like a cleaner solution to #984 

Example:
```nix
pkgs.haskell-nix.project' {
  src = ./.;
  compiler-nix-name = "ghcXYZ";
  pkgs-mappers =
  {
    X11 = pkgs.buildEnv {
      name = "x11-haskell-wrapper";
      paths = with pkgs.xorg; [ libX11 libXScrnSaver libXinerama ]
        ++ [ pkgs.xlibs.libXinerama.dev ];
    };
  };
}
```

I'll add a test to `ci.nix` once this PR gets it's initial okay.

rvl commented:
Thanks for this @paumr. It's not quite what I had in mind with https://github.com/input-output-hk/haskell.nix/pull/984#discussion_r550837042 though.

Overlays would be completely adequate for users who need to configure their own mappings, so I don't think there is a need for configurable pkgs-mappers.

However, building an env for "composite" system dependencies is definitely something that we want.

I have pushed a commit to this PR branch - this lets a list of packages be used in the mapping. I think it should work but haven't tested it for your xmonad use case. Please let me know what you think.

angerman commented:
This sounds vaguely familiar. @hamishmack does this ring a bell? Did we discuss this at some point? Do you remember anything?

paumr commented:
> Thanks for this @paumr. It's not quite what I had in mind with [#984 (comment)](https://github.com/input-output-hk/haskell.nix/pull/984#discussion_r550837042) though.
> 
> Overlays would be completely adequate for users who need to configure their own mappings, so I don't think there is a need for configurable pkgs-mappers.
> 
> However, building an env for "composite" system dependencies is definitely something that we want.
> 
> I have pushed a commit to this PR branch - this lets a list of packages be used in the mapping. I think it should work but haven't tested it for your xmonad use case. Please let me know what you think.

I see what you mean, thanks!
After grepping through the project it seems like `pkgconfPkgs` isn't actually used, maybe we could remove it?

michaelpj commented:
> After grepping through the project it seems like pkgconfPkgs isn't actually used, maybe we could remove it?

:scream_cat:

rvl commented:
I thought about it some more and
1. For any mapping that a user supplies, using an overlay is nicer.
2. The default mapping needs to go _underneath_ anything that the user provides.
3. Perhaps using a module option rather than a mkPkgSet function parameter is a better way to configure this (may be difficult to implement though).

So @paumr  - do as you wish with the commits I added - improve or revert.

@michaelpj, @hamishmack , what do you think about the above points?

hamishmack commented:
The correct fix might be to add `Xss` and `Xinerama` as `extra-libraries` in X11.cabal [here]( https://github.com/xmonad/X11/blob/93f8ba305d72dfe44c440aba05dfecbeb1d7d74a/X11.cabal#L74).  They could be guarded by cabal flags if they are not always required.

I have added a mapping for `Xss = [ xorg.libXScrnSaver ];`.

I could not find when the `.dev` of `libXinerama` is needed, just including `xorg.libXinerama` seemed to be enough for the X11 library component to build with the `configure` output indicating `Xinerama` was found (tested with nixpkgs 20.09 and 20.03). 
 Is there something that breaks?  I also could not find a difference between `xorg` and `xlibs` versions of Xinerama, the hashes for them seem identical.

I don't like `system-pkgs-overlay`.  Both these work for me with latest master and I think I prefer this approach to working around issues like this:

```nix
pkgs.haskell-nix.tool "ghc8102" "xmonad" {
  version = "latest";
  modules = [{
    # Add missing `extra-libraries` dependencies
    packages.X11.components.library.libs = [ pkgs.xorg.libXScrnSaver pkgs.xorg.libXinerama ];
  }];
}
```

```nix
pkgs.haskell-nix.tool "ghc8102" "xmonad" {
  version = "latest";
  modules = [{
    # Replace `extra-libraries` dependencies
    packages.X11.components.library.libs = pkgs.lib.mkForce (with pkgs.xorg;
      [ libX11 libXrandr libXext libXScrnSaver libXinerama ]);
  }];
}
```

I have added a `defaultModule` for `packages.X11.components.library.libs` in a new `overlays/default-modules.nix`.  It is unlikely someone would have a `X11` package that would break as result of the extra dependencies.

To try it out you can run:

```
nix-build -E '((import ./. {}).pkgs.haskell-nix.tool "ghc8102" "xmonad" "latest")'
```

hamishmack commented on lib/system-nixpkgs-map.nix:
```diff
@@ -1,84 +1,84 @@
-# cabal file to nixpkgs pkgs.
-# See ../docs/dev/pkg-map.md
+# Maps of system pkg references from cabal file to nixpkgs pkgs.
+# haskell lib name -> nixpkgs package(s) mapping
```
It's not really a `haskell lib name`.   How about:

```
# library name (from `extra-libraries` field of `.cabal` file) -> nixpkgs package(s) mapping.
# This mapping is used to construct the `libs` component attribute.
```

michaelpj commented:
> a new overlays/default-modules.nix

Hmmm, does this perhaps subsume (or improve on) `hackage-quirks`?

michaelpj commented:
> For any mapping that a user supplies, using an overlay is nicer.

I've actually changed my mind about this. Overlays potentially change other things. For example, maybe cabal uses name N for library X whereas nixpkgs uses it for library Y. If you made an overlay mapping N to X, you might break lots of things in nixpkgs. That is, there's no guarantee that then names cabal wants will be free. This really is about how to "adapt" cabal's names to nixpkgs names, so it does make sense to have a separate thing for it.

> Perhaps using a module option rather than a mkPkgSet function parameter is a better way to configure this

Module options let us do merging, which is nice, and also

> The default mapping needs to go underneath anything that the user provides.

With suitable usage of `mkDefault` and friends we might be able to do this relatively straightforwardly.

> The correct fix might be to add Xss and Xinerama as extra-libraries in X11.cabal here. They could be guarded by cabal flags if they are not always required.

I agree, there does seem like there's an upstream issue here that ought to be fixed there.

hamishmack commented:
> Hmmm, does this perhaps subsume (or improve on) hackage-quirks?

I think the general rule is probably use `hackage-quirks.nix` for executables and `default-modules.nix` for libraries (when possible).  One big difference is that `default-modules.nix` can only change things in the module system for components where as `hackage-quirks.nix` can also change the other `project` arguments (things that affect the generation of the plan).

hamishmack commented:
Long term I think we should change the way `extral-libraries` are resolved.  Using the `pkgs` argument for this seems wrong (since the values in `extra-libraries` are not package names).  It would be nicer to have something more like the way `pkgconf-depends` are resolved through the use of a separate `pkgconfPkgs` argument (yes it is used but by the output of `plan-to-nix` so grep is not obvious).

To do this we would first need to update `plan-to-nix` so that it output `systemPkgs.X11` or `libPkgs.X11` instead of `pkgs.X11` for `extra-libraries: X11`.  One advantage would be we would not need to use the `buildEnv` trick to enable multiple derivations to be included for a library (the `pkgconf-depends` just uses `builtins.concatLists` instead).

I think we should probably move `nix-tools` into `haskell.nix` first though.  There might be a lot of materialization updates and it would be nice not to have to deal with multiple repos.

paumr commented:
@rvl:
> For any mapping that a user supplies, using an overlay is nicer.

I'm on the fence about using overlays, initially i decided against them - sharing @michaelpj 's reasoning ( https://github.com/input-output-hk/haskell.nix/pull/988#issuecomment-754535202 ).
But on the other hand overlays are the customary procedure of modification in nix.
> The default mapping needs to go underneath anything that the user provides.

If by "underneath" you mean "overwriteable by the user" i agree.

> So @paumr - do as you wish with the commits I added - improve or revert.

I'll try to help, but since this codebase is quite new to me and since this boils down to a design-desicion i doubt that i'll revert anything from you guys.

@hamishmack:
> The correct fix might be to add Xss and Xinerama as extra-libraries in X11.cabal here. They could be guarded by cabal flags if they are not always required.

`Xss` seems to be always required since https://github.com/xmonad/X11/pull/61
Adding them to extra-libraries sounds like a decent solution!

> I have added a mapping for Xss = [ xorg.libXScrnSaver ];.

this caused me to add `Xrender = [ xorg.libXrender ];`.
Should changes like that be factored out and split into multiple PRs upon cleanup? Or will all just be merged together?

> I could not find when the .dev of libXinerama is needed, just including xorg.libXinerama seemed to be enough

It is, only the `.dev` version seems to include `Xinerama.h`:

```
 > ~/playground/haskell.nix nix-locate Xinerama.h
haskellPackages.X11.doc                          23,886 r /nix/store/rs46azn5a7vb6m5ac08ahnr914kx0dcn-X11-1.9-doc/share/doc/X11-1.9/html/Graphics-X11-Xinerama.html
haskellPackages.X11.doc                          41,128 r /nix/store/rs46azn5a7vb6m5ac08ahnr914kx0dcn-X11-1.9-doc/share/doc/X11-1.9/html/src/Graphics.X11.Xinerama.html
haskellPackages.X11.out                          29,407 r /nix/store/87mwnfpzrwmclf6aqc1aakq88l1hyff8-X11-1.9/lib/ghc-8.6.4/x86_64-linux-ghc-8.6.4/X11-1.9-8yf1oSectWR6hs7SaASgaB/Graphics/X11/Xinerama.hi
(xmonad-with-packages.out)                            0 s /nix/store/jdr8x1y301kqs4w1zvfwj5yq0yw2z5vv-ghc-8.6.4-with-packages/lib/ghc-8.6.4/x86_64-linux-ghc-8.6.4/X11-1.9-8yf1oSectWR6hs7SaASgaB/Graphics/X11/Xinerama.hi
(winePackages.unstable.out)                       1,964 r /nix/store/hakgk9fir7ddlh90v50r2pfq71qbycqx-libXinerama-1.1.4-dev/include/X11/extensions/Xinerama.h
(taffybar.out)                                        0 s /nix/store/dgpirzvwl6yl52cfbm44pa4sgnxigmb8-ghc-8.6.4-with-packages/lib/ghc-8.6.4/x86_64-linux-ghc-8.6.4/X11-1.9-8yf1oSectWR6hs7SaASgaB/Graphics/X11/Xinerama.hi
xlibs.libXinerama.dev                             1,964 r /nix/store/ib61jgwqxxiqvivjfsh3j2cdz0hrrzgf-libXinerama-1.1.4-dev/include/X11/extensions/Xinerama.h
```

**Noteworthy**:
if you build: `libXinerama.dev` both `libXinerama.dev` and `libXinerama` are the build product.
if you build: `libXinerama` only `libXinerama` will be the build product.
if you build: `libXinerama.dev` **and** if you remove both symlinks (to `libXinerama` and `libXinerama.dev`) **and** if you build `libXinerama` again, `libXinerama.dev` and `libXinerama` will be the build product.

Hopefully that's a local issue, being strict enough to create the link but too lazy to actually build the package doesn't look too good. (hope i'm not too strict :P)


> Is there something that breaks? I also could not find a difference between xorg and xlibs versions of Xinerama, the hashes for them seem identical.

You are right xlibs is aliased to xorg! https://github.com/NixOS/nixpkgs/blob/d9dba88d08a9cdf483c3d45f0d7220cf97a4ce64/pkgs/top-level/aliases.nix#L711

> I have added a defaultModule for packages.X11.components.library.libs in a new overlays/default-modules.nix. It is unlikely someone would have a X11 package that would break as result of the extra dependencies.

Means you are proposing to replace `system-pkgs-overlay` with `overlays/default-modules.nix`?

hamishmack commented:
> Means you are proposing to replace system-pkgs-overlay with overlays/default-modules.nix?

Until we find a situation that requires it. I think most use cases will be covered by one of:

* Adding an overlay to nixpkgs itself where we want to change the package itself (for instance we use that to [turn on Gtk debugging](https://github.com/leksah/leksah/blob/master/nix/overlays/gtk-debug.nix) for Leksah).

* Updating the system-nixpkgs-map.nix to add missing libraries.

* Using `modules` and add to or replace some `packages.*.components.*.libs`.

* Adding `defaultModules` to `haskell.nix` for cases like the `X11` package.

The system-pkgs-overlay arg was a nice idea and well implemented, but I think we can avoid the extra complexity for now.

> Should changes like that be factored out and split into multiple PRs upon cleanup? Or will all just be merged together?

I think the PR is not to big merge together.  I'll check it with bors tomorrow and merge it unless you can think of anything else we should add.  We should rename it to something like "Updated system-nixpkgs-map.nix and a default module for X11 package (fixes #984)".

paumr commented:
I added a sample based on https://github.com/input-output-hk/haskell.nix/pull/988#issuecomment-754400500 to the docs. Other than that merging should be fine by me.

Thanks for the help and for adding this feature!

michaelpj commented on overlays/default-modules.nix:
```diff
@@ -0,0 +1,22 @@
+# This overlay adds `defaultModules` to work around issues with
```
Further random poking around: is this not the same in function as `modules/configuration-nix.nix`?

hamishmack commented:
bors try
